it/common/src/main/java/org/apache/beam/it/common/PipelineLauncher.java

```diff
@@ -0,0 +1,422 @@
+
+    private final String text;
+
+    JobState(String text) {
+      this.text = text;
+    }
+
+    /**
+     * Parses the state from Dataflow.
+     *
+     * <p>Always use this in place of valueOf.
+     */
+    public static JobState parse(String fromDataflow) {
+      return valueOf(fromDataflow.replace(DATAFLOW_PREFIX, ""));
+    }
+
+    @Override
+    public String toString() {
+      return text;
+    }
+  }
+
+  /** Config for starting a Dataflow job. */
+  class LaunchConfig {
+    private final String jobName;
+    private final ImmutableMap<String, String> parameters;
+    private final ImmutableMap<String, Object> environment;
+    @Nullable private final String specPath;
+    @Nullable private final Sdk sdk;
+    @Nullable private final String executable;
+    @Nullable private final Pipeline pipeline;
+
+    private LaunchConfig(Builder builder) {
+      this.jobName = builder.jobName;
+      this.parameters = ImmutableMap.copyOf(builder.parameters);
+      this.environment = ImmutableMap.copyOf(builder.environment);
+      this.specPath = builder.specPath;
+      this.sdk = builder.sdk;
+      this.executable = builder.executable;
+      this.pipeline = builder.pipeline;
+    }
+
+    public String jobName() {
+      return jobName;
+    }
+
+    public ImmutableMap<String, String> parameters() {
+      return parameters;
+    }
+
+    public ImmutableMap<String, Object> environment() {
+      return environment;
+    }
+
+    @Nullable
+    public String getParameter(String key) {
+      return parameters.get(key);
+    }
+
+    public @Nullable String specPath() {
+      return specPath;
+    }
+
+    public @Nullable Sdk sdk() {
+      return sdk;
+    }
+
+    public @Nullable String executable() {
+      return executable;
+    }
+
+    public @Nullable Pipeline pipeline() {
+      return pipeline;
+    }
+
+    public static Builder builderWithName(String jobName, String specPath) {
+      return new Builder(jobName, specPath);
+    }
+
+    public static Builder builder(String testName, @Nullable String specPath) {
+      return new Builder(createJobName(testName), specPath);
+    }
+
+    public static Builder builder(String jobName) {
+      return builder(jobName, null);
+    }
+
+    /** Builder for the {@link LaunchConfig}. */
+    public static final class Builder {
+      private final String jobName;
+      private final @Nullable String specPath;
+      private final Map<String, Object> environment;
+      private Map<String, String> parameters;
+      private @Nullable Sdk sdk;
+      private @Nullable String executable;
+      private @Nullable Pipeline pipeline;
+
+      private Builder(String jobName, @Nullable String specPath) {
+        this.jobName = jobName;
+        this.parameters = new HashMap<>();
+        this.environment = new HashMap<>();
+        this.specPath = specPath;
+      }
+
+      public String getJobName() {
+        return jobName;
+      }
+
+      @Nullable
```

Comment:
Use checkerframework `@Nullable` annotation and annotate the return type, not the overall method.

Do this throughout the file, and all the new code.